src/common/Spine.ts

```diff
@@ -178,13 +190,19 @@ export class Spine<N extends IR.Node = IR.Node> {
   }
 }
 
-export type PluginVisitor<T> = <N extends IR.Node>(
+export type PluginVisitor<T = IR.Node[] | IR.Node | undefined> = <
+  N extends IR.Node,
+>(
   node: N,
   spine: Spine<N>,
   context: CompilationContext,
 ) => T;
 
-export type Visitor<T> = <N extends IR.Node>(node: N, spine: Spine<N>) => T;
+export type Visitor<T> = <N extends IR.Node>(
+  node: N,
+  spine: Spine<N>,
+  skip: SkipOverride,
```

Comment:
Instead of passing `skip: SkipOverride`, it would be safer (in case of a future adding more functions like `skip`) to pass in `ctx: VisitorContext` where
```ts
interface VisitorContext {
  skip: SkipOverride
}
interface CompilationContext extends VisitorContext {
  options: CompilationOptions;
  addWarning: AddWarning;
}
```
It could even be
```ts
interface VisitorContext {
  skipReplacement: () => void;
  skipChildren: () => void;
}
interface CompilationContext extends VisitorContext {
  options: CompilationOptions;
  addWarning: AddWarning;
}
```

<!--This could be more aligned with `PluginVisitor` (and be future-proofed for adding other functions like `skip`) if the `skip` function is part of a `VisitorContext` object. So `VisitorContext` has `skip` (or `skipReplacement` and `skipChildren`), then `CompilationContext` extends `VisitorContext` with its extra options. This would allow a plugin to receive both a compilation context and a visitor context. -->

Reply:
Nice